File: frontend/src/types/contract.ts
import { z } from "zod/v4"

export const contractSchema = z.object({
  id: z.string().uuid(),
  categoryId: z.string().uuid(),
  name: z.string().min(1),
  productName: z.string().optional(),
  company: z.string().optional(),
  contractNumber: z.string().optional(),
  customerNumber: z.string().optional(),
  pricePerMonth: z.number().nonnegative().optional(),
  startDate: z.string().date(),
  endDate: z.string().date().optional(),
  minimumDurationMonths: z.number().int().nonnegative(),
  extensionDurationMonths: z.number().int().nonnegative(),
  noticePeriodMonths: z.number().int().nonnegative(),
  customerPortalUrl: z.string().url().optional().or(z.literal("")),
  paperlessUrl: z.string().url().optional().or(z.literal("")),
  comments: z.string().optional(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
})

export type Contract = z.infer<typeof contractSchema>

export const contractFormSchema = z.object({
  name: z.string().min(1),
  productName: z.string().optional(),
  company: z.string().optional(),
  contractNumber: z.string().optional(),
  customerNumber: z.string().optional(),
  pricePerMonth: z.number().nonnegative().optional(),
  startDate: z.string().date(),
  endDate: z.string().date().optional(),
  minimumDurationMonths: z.number().int().nonnegative(),
  extensionDurationMonths: z.number().int().nonnegative(),
  noticePeriodMonths: z.number().int().nonnegative(),
  customerPortalUrl: z.string().url().optional().or(z.literal("")),
  paperlessUrl: z.string().url().optional().or(z.literal("")),
  comments: z.string().optional(),
})

export type ContractFormData = z.infer<typeof contractFormSchema>

File: backend/internal/model/contract.go
package model

import (
	"errors"
	"time"

	"github.com/google/uuid"
)

type Contract struct {
	ID                      uuid.UUID `json:"id"`
	CategoryID              uuid.UUID `json:"categoryId"`
	Name                    string    `json:"name"`
	ProductName             string    `json:"productName,omitempty"`
	Company                 string    `json:"company,omitempty"`
	ContractNumber          string    `json:"contractNumber,omitempty"`
	CustomerNumber          string    `json:"customerNumber,omitempty"`
	PricePerMonth           *float64  `json:"pricePerMonth,omitempty"`
	StartDate               string    `json:"startDate"`
	EndDate                 string    `json:"endDate,omitempty"`
	MinimumDurationMonths   int       `json:"minimumDurationMonths"`
	ExtensionDurationMonths int       `json:"extensionDurationMonths"`
	NoticePeriodMonths      int       `json:"noticePeriodMonths"`
	CustomerPortalURL       string    `json:"customerPortalUrl,omitempty"`
	PaperlessURL            string    `json:"paperlessUrl,omitempty"`
	Comments                string    `json:"comments,omitempty"`
	CreatedAt               time.Time `json:"createdAt"`
	UpdatedAt               time.Time `json:"updatedAt"`
}

type ContractInput struct {
	Name                    string   `json:"name"`
	ProductName             string   `json:"productName,omitempty"`
	Company                 string   `json:"company,omitempty"`
	ContractNumber          string   `json:"contractNumber,omitempty"`
	CustomerNumber          string   `json:"customerNumber,omitempty"`
	PricePerMonth           *float64 `json:"pricePerMonth,omitempty"`
	StartDate               string   `json:"startDate"`
	EndDate                 string   `json:"endDate,omitempty"`
	MinimumDurationMonths   int      `json:"minimumDurationMonths"`
	ExtensionDurationMonths int      `json:"extensionDurationMonths"`
	NoticePeriodMonths      int      `json:"noticePeriodMonths"`
	CustomerPortalURL       string   `json:"customerPortalUrl,omitempty"`
	PaperlessURL            string   `json:"paperlessUrl,omitempty"`
	Comments                string   `json:"comments,omitempty"`
}

func (c *ContractInput) Validate() error {
	if c.Name == "" {
		return errors.New("name is required")
	}
	if c.StartDate == "" {
		return errors.New("startDate is required")
	}
	return nil
}
